R/generate_dust.R

```diff
@@ -448,20 +438,23 @@ generate_dust_lhs <- function(lhs, dat, name_state, options) {
       generate_dust_sexp(name, dat$sexp_data, options)
     }
   } else if (location == "state") {
+    packing <- dat$storage$packing$state
+    offset <- packing$offset[[match(name, packing$name)]]
     if (is_array) {
-      stop("support arrays in variables")
+      offset <- expr_plus(idx, offset)
     }
-    ## TODO: this wil need to change, once we support arrays: at that
-    ## point we'll make this more efficient too by computing
-    ## expressions for access.
     sprintf("%s[%s]",
-            name_state, match(name, dat$storage$packing$state$scalar) - 1)
+            name_state,
+            generate_dust_sexp(offset, dat$sexp_data, options))
   } else if (location == "adjoint") {
+    packing <- dat$storage$packing$adjoint
+    offset <- packing$offset[[match(name, packing$name)]]
     if (is_array) {
-      stop("support arrays in adjoint")
+      offset <- expr_plus(idx, offset)
```

Comment:
Not covered

Reply:
```
test_that("expr_plus", {
  expect_identical(expr_plus(1, 2), 3)
  x <- expr_plus(quote(var), 2)
  expect_type(x, "language")
  x <- as.character(x)
  expect_equal(x, c("+", "var", "2"))
})
```

Reply:
this one here I think needs to wait as we prevent use of differentiation in systems with arrays - I've put an assertion in now above the uncovered line to mark that